### ISSUE TITLE
Failed to launch ptxas error when using nvidia/cuda runtime Docker image

### ISSUE THREAD

**Issue by goncinious** [tensorflow/tensorflow#60737]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Have you reproduced the bug with TF nightly?

Yes

### Source

source

### Tensorflow Version

2.14.0-dev20230531

### Custom Code

Yes

### OS Platform and Distribution

20.04

### Python version

3.9.16

### CUDA/cuDNN version

CUDA 11.8.0 / cuDNN 8.7

### Current Behaviour?

I get the following warning when I run prediction using a 3D model (full log provided below):
```
2023-05-31 11:06:53.982167: W tensorflow/compiler/xla/stream_executor/gpu/redzone_allocator.cc:318] INTERNAL: Failed to launch ptxas
Relying on driver to perform ptx compilation. 
Modify $PATH to customize ptxas location.
This message will be only logged once.
```
I'm using a model that takes a 3D input and the error occurs when I have a `Conv3DTranspose` layer (which my model implements).

I'm using the official [`nvidia/cuda`](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/cuda) image with runtime libraries, specifically the `nvcr.io/nvidia/cuda:11.8.0-cudnn8-runtime-ubuntu20.0`.

I don't get the warning when using the `devel` image, however, I want to stick with `runtime`, as it's much smaller and I don't need any of the compiling capabilities or dev tools.

Does this warning have an impact on model inference and how can I fix it?

Minimal Dockerfile and Python script to reproduce issue are provided below.

### Standalone code to reproduce the issue

Dockerfile:
```shell

FROM nvcr.io/nvidia/cuda:11.8.0-cudnn8-runtime-ubuntu20.04

ENV PYENV_ROOT=/root/.pyenv \
    PATH="/root/.local/bin:/opt/venv/bin:/root/.pyenv/shims/:root/.pyenv/bin:${PATH}" \
    DEBIAN_FRONTEND=noninteractive

ARG PYTHON_VERSION=3.9

RUN apt update -y && apt upgrade -y && \
    apt-get install --no-install-recommends -y wget build-essential curl git && \
    # install python via pyenv
    apt-get install -y --no-install-recommends libreadline-gplv2-dev libncursesw5-dev libssl-dev libsqlite3-dev tk-dev libgdbm-dev libc6-dev libbz2-dev libffi-dev zlib1g-dev liblzma-dev && \
    curl https://pyenv.run | bash && \
    pyenv update && \
    pyenv install ${PYTHON_VERSION} && \
    pyenv global ${PYTHON_VERSION} && \
    pyenv rehash && \
    # install tensorflow
    python${PYTHON_MAJOR_VERSION} -m pip install tf-nightly
```

`test.py`:
```python
import numpy as np
import tensorflow as tf
from tensorflow.keras import Input
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Conv3DTranspose

def build_model(input_shape):
    inputs = Input(shape=input_shape)
    x = Conv3DTranspose(filters=1, kernel_size=(1, 1, 1), strides=(1, 1, 1))(inputs)
    return Model(inputs=inputs, outputs=x)

print(tf.version.GIT_VERSION, tf.version.VERSION)

input_shape = (10, 10, 10, 1)
model = build_model(input_shape)

# input of shape (5, 10, 10, 10, 1)
test_input = np.ones(shape=(5, *input_shape), dtype=np.float32)

output = model.predict(test_input, verbose=1, batch_size=1)
```

Steps to reproduce:
```
docker build -t tf-issue -f /path/to/Dockerfile .
docker run --gpus all -v /path/to/test.py:/test.py tf-issue python /test.py
```

### Relevant log output

```shell
==========
== CUDA ==
==========

CUDA Version 11.8.0

Container image Copyright (c) 2016-2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.

This container image and its contents are governed by the NVIDIA Deep Learning Container License.
By pulling and using the container, you accept the terms and conditions of this license:
https://developer.nvidia.com/ngc/nvidia-deep-learning-container-license

A copy of this license is made available in this container at /NGC-DL-CONTAINER-LICENSE for your convenience.

2023-05-31 11:06:49.787277: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:7704] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2023-05-31 11:06:49.787340: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2023-05-31 11:06:49.787352: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1520] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2023-05-31 11:06:49.794051: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-05-31 11:06:50.626364: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
2023-05-31 11:06:51.483784: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:995] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2023-05-31 11:06:51.504146: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:995] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2023-05-31 11:06:51.504472: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:995] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2023-05-31 11:06:51.505671: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:995] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2023-05-31 11:06:51.505938: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:995] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2023-05-31 11:06:51.506189: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:995] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2023-05-31 11:06:52.004673: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:995] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2023-05-31 11:06:52.004929: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:995] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2023-05-31 11:06:52.005143: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:995] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2023-05-31 11:06:52.005338: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1639] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 20638 MB memory:  -> device: 0, name: NVIDIA A10G, pci bus id: 0000:00:1e.0, compute capability: 8.6
2023-05-31 11:06:53.285968: I tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:434] Loaded cuDNN version 8700
2023-05-31 11:06:53.981061: I tensorflow/tsl/platform/default/subprocess.cc:304] Start cannot spawn child process: No such file or directory
2023-05-31 11:06:53.981512: I tensorflow/tsl/platform/default/subprocess.cc:304] Start cannot spawn child process: No such file or directory
2023-05-31 11:06:53.981543: W tensorflow/compiler/xla/stream_executor/gpu/asm_compiler.cc:109] Couldn't get ptxas version : FAILED_PRECONDITION: Couldn't get ptxas/nvlink version string: INTERNAL: Couldn't invoke ptxas --version
2023-05-31 11:06:53.982117: I tensorflow/tsl/platform/default/subprocess.cc:304] Start cannot spawn child process: No such file or directory
2023-05-31 11:06:53.982167: W tensorflow/compiler/xla/stream_executor/gpu/redzone_allocator.cc:318] INTERNAL: Failed to launch ptxas
Relying on driver to perform ptx compilation. 
Modify $PATH to customize ptxas location.
This message will be only logged once.
v1.12.1-94762-gf8066222ad6 2.14.0-dev20230531
5/5 [==============================] - 2s 1ms/step
```
</details>

**Comment by SuryanarayanaY** [comment]
Hi @goncinious ,

Could you please try installing cuda-nvcc package using `conda install -c "nvidia/label/cuda-11.8.0" cuda-nvcc` and let us know if it helps.

Thanks!

**Comment by goncinious** [comment]
Thanks @SuryanarayanaY! I'm already using [Poetry](https://python-poetry.org/) for managing dependencies in my use case, so I'd prefer to avoid having to install `conda` just for this. Can I install `cuda-nvcc` without `conda`?

**Comment by SuryanarayanaY** [comment]
Hi @goncinious ,



> Yes, I believe the same applies to the `devel` image from the official `nvidia/cuda` image. As mentioned above, it doesn't show the warning highlighted above. However, the point here is how to properly setup `ptxas` compilation while avoiding installing massive dependencies in the `runtime` image.

AFAIK ptxas is a package that included in CUDA package itself and I am not sure this falls under TF scope.May be we need to confirm the same with Nvidia team. I believe devel images dont have this problem with ptxas path as per reference from the attached issues from Jax, [link1](https://github.com/google/jax/discussions/6843) and [link2](https://github.com/google/jax/discussions/10327).



> 2023-06-19 11:12:26.647185: E tensorflow/compiler/xla/stream_executor/gpu/asm_compiler.cc:114] *** WARNING *** You are using ptxas 10.1.243, which is older than 11.1. ptxas before 11.1 is known to miscompile XLA code, leading to incorrect results or invalid-address errors.

I can see this as logged as Error log may be due to the reason that it might miscompile XLA code.

**Comment by SuryanarayanaY** [comment]
@goncinious ,

As per Official TF documentation for Docker support you might need to install Nvidia driver in host system and Nvidia container toolkit on docker, the instructions for which mentioned [here](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html#docker).

Could you please check whether the above resources solves your problem

Thanks!

**Comment by kevinthegreat1** [comment]
I followed this guide <https://www.tensorflow.org/install/pip#windows-wsl2> and had the same issue where jupyter notebook kernel was crashing because it couldn't find ptxas. `conda install -c "nvidia/label/cuda-11.8.0" cuda-nvcc` fixed the issue. Thanks.
